Fix links to internal documents

rw1nkler opened the pull request:
The links to markdown files used by the Sphinx documentation are not working correctly when using the Markdown Symlinks locally. 

This PR introduces the changes that allow generating correct documentation locally. The changes need to be tested on RTD. 
The PR also provides basic tests for the Sphinx extension.

mithro commented:
@acomodi Was the last person to look at this, please get him to review it first.

rw1nkler committed:
Add markdown-code-symlinks tests

This PR adds a simple test for a markdown-code-symlinks extension.
Test check whether the links where correctly substituted by the extension,
using Travis CI.

Signed-off-by: Robert Winkler <rwinkler@antmicro.com>

rw1nkler committed:
Add config that allows building tests on RTD

Signed-off-by: Robert Winkler <rwinkler@antmicro.com>

rw1nkler committed:
Fix markdown-code-symlinks for Sphinx >= 3.0

Signed-off-by: Robert Winkler <rwinkler@antmicro.com>

rw1nkler commented:
The problem with my local build was caused by Sphinx >= 3.0.  Because the file extensions have been removed from the Sphinx AST nodes, the extension was producing wrong links. The changes introduced in this PR resolve the problem.

Additionally, I created a simple Travis tests that will help to track the problems with this extension.

Currently, two tests are commented out, because they are failing. When the linked markdown file contains a link to another markdown file that is not included in the Sphinx toctree, the link cannot be correctly resolved. At this moment, I don't know how to resolve this issue because the problem requires tracking the symbolic link target. Currently, I cannot find any information in Sphinx AST nodes about the source file, which contains the link that has to be resolved.

acomodi commented on markdown_code_symlinks/__init__.py:
```diff
@@ -106,20 +108,45 @@ def relative_docs(cls, url):
 Current Value: {}
 """.format(docs_rel, cls.mapping['docs2code'][docs_rel])
+            assert code_rel_no_ext not in cls.mapping['code2docs'], """\
+Assertion error! Document already in mapping!
+    New Value: {}
+Current Value: {}
+""".format(docs_rel, cls.mapping['code2docs'][code_rel])
```
There are at least four of these prints. Probably it would be good to wrap this string formatting somewhere, maybe by defining the string beforehand and then formatting it when needed.

rw1nkler replied:
Fixed

acomodi reviewed (changes requested):
Generally LGTM, just a small fix IMO.

acomodi commented on markdown_code_symlinks/__init__.py:
```diff
@@ -106,20 +108,31 @@ def relative_docs(cls, url):
-Current Value: {}
-""".format(docs_rel, cls.mapping['code2docs'][code_rel])
+        if sphinx_v < sphinx_v3_0:
+            if docs_rel in cls.mapping['docs2code']:
+                assert False, assert_msg_temp.format(docs_rel, cls.mapping['docs2code'][docs_rel])
```
I would remove the if statement and leave the assert

```suggestion
            assert code_rel not in cls.mapping['code2docs'], assert_msg_temp.format(docs_rel, cls.mapping['docs2code'][docs_rel])
```

rw1nkler replied:
Fixed

acomodi reviewed (approved):
LGTM
